### PR TITLE
Promote local memory to SSA registers

### LINKED ISSUE #2770: Write a `mem2reg` optimisation pass
For code such as

```
let a: u8 = 2;                                                                                                                                                                                      
let b: u8 = 22;                                                                                                                                                                                     
assert(__add(a, b) == 24);
```
we generate the IR

```
        local ptr u64 a
        local ptr u64 a_
        local ptr u64 a__
        local ptr u64 a___
        local ptr u64 b
        local ptr u64 b_
        local ptr u64 b__
        local ptr u64 b___

        entry:
        v0 = get_ptr ptr u64 a, ptr u64, 0, !2
        v1 = const u64 2, !3
        store v1, ptr v0, !2
        v2 = get_ptr ptr u64 b, ptr u64, 0, !4
        v3 = const u64 22, !5
        store v3, ptr v2, !4
        v4 = get_ptr ptr u64 a, ptr u64, 0, !6
        v5 = load ptr v4, !6
        v6 = get_ptr ptr u64 b, ptr u64, 0, !7
        v7 = load ptr v6, !7
        v8 = add v5 v7
        v9 = const u64 24, !8
        v10 = cmp eq v8 v9, !9
        v11 = const bool false, !11
        v12 = cmp eq v10 v11, !15
        cbr v12, assert_0_block0, assert_0_block1, !16
```
A [mem2reg](https://llvm.org/docs/Passes.html#mem2reg-promote-memory-to-register) pass will promote these memory variables to live in SSA registers, thus saving on a bunch of loads and stores.

### PR DESCRIPTION
**Pull request by vaivaswatha**
Closes #2770. 

Benchmark applications: https://github.com/FuelLabs/sway-applications

-----------------------------------
benchmark | before | after
----------------|----------|--------
dao-voting | 25784 | 20568
escrow | 95076 | 72740
fundraiser | 41816 | 31344
multisig-wallet | 8112 | 6592
NFT | 28552 | 23200
oracle | 2132 | 1924

### CODE REVIEW

File: sway-ir/src/optimize/dominator.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+                Some(idom) if idom == new_idom => {}
+                _ => {
+                    b_node.parent = Some(new_idom);
+                    changed = true;
+                }
+            }
+        }
+    }
+
+    fn intersect(
```
Comment (mohammadfawaz) [inline sway-ir/src/optimize/dominator.rs:124]: Maybe a comment explaining what `intersect` does.

**Review by mohammadfawaz — COMMENTED** [review]
This is very exciting, thanks @vaivaswatha. Just some nits.

File: sway-ir/src/optimize/mem2reg.rs [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+pub fn promote_to_registers(context: &mut Context, function: &Function) -> Result<bool, IrError> {
+    let safe_locals = filter_usable_locals(context, function);
+
+    if safe_locals.is_empty() {
+        return Ok(false);
+    }
+
+    let (dom_tree, po) = compute_dom_tree(context, function);
+    let dom_fronts = compute_dom_fronts(context, &dom_tree);
+    // print!(
```
[the remark this answers is not in the document]
Comment (vaivaswatha) [reply sway-ir/src/optimize/mem2reg.rs:142]: > Do we need this commented `print!`?

It should ideally be something like LLVM's `DEBUG()` macro, but I'm not sure if we have that. I'm confused as to what's better. delete it or have it as a comment for future debugging.

File: sway-ir/src/optimize/dominator.rs (hunk 1 above)
Comment (vaivaswatha) [reply sway-ir/src/optimize/dominator.rs:124]: > Maybe a comment explaining what `intersect` does.

It's a bit hard to describe what it does without the context of the paper. Let me give it a try !

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/alloc/src/main.sw [hunk 3]
```diff
@@ -7,9 +7,9 @@ use std::context::registers::*;
 use std::assert::assert;
 
 fn lw(ptr: raw_ptr) -> u64 {
-    asm(r1: ptr) {
-        lw r1 r1 i0;
-        r1: u64
+    asm(r1: ptr, res) {
```
[the remark this answers is not in the document]
Comment (mohammadfawaz) [reply test/src/e2e_vm_tests/test_programs/should_pass/stdlib/alloc/src/main.sw:10]: I'm ok with disallowing `asm` blocks to read and write the same register temporarily. Should be a simple check in the parser or type checking. I can create an issue.

File: sway-ir/src/optimize/dominator.rs [hunk 4]
```diff
@@ -0,0 +1,209 @@
+use crate::{block::Block, BranchToWithArgs, Context, Function};
```
[the remark this answers is not in the document]
Comment (mohammadfawaz) [reply sway-ir/src/optimize/dominator.rs:1]: Make sense to me.

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/alloc/src/main.sw (hunk 3 above)
[the remark this answers is not in the document]
Comment (vaivaswatha) [reply test/src/e2e_vm_tests/test_programs/should_pass/stdlib/alloc/src/main.sw:10]: That makes sense, thank you.

File: sway-ir/src/optimize/mem2reg.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mohammadfawaz) [reply sway-ir/src/optimize/mem2reg.rs:142]: Yes! I've proposed this before here: https://github.com/FuelLabs/sway/issues/2242

My personal preference is not to check in commented code, but totally up to you :)

**Review by mohammadfawaz — APPROVED** [review]

**Review by mohammadfawaz — APPROVED** [review]

**Review by sezna — APPROVED** [review]

**Pull request merged**